Add pristine options to "west build" and the config

carlescufi opened the pull request:
See commit messages for additional info.

carlescufi committed:
west: build: Use officially documented CMake options

The -B option has always existed but was first officially documented in
CMake 3.13.0. In that same release the -S option was introduced,
replacing the old undocumented -H. Switch to using the officially
documented options.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
west: build: Add new pristine cmd-line and config option

Add a new command-line and build config option, `pristine`, that the
user can pass to `west build` or set in its configuration file(s) in
order to automatically trigger a pristine build on every build or
whenever west considers it required.

The option can take the following values:

- never: Never run the target
- always: Always run the pristine target before building
- auto: Run the pristine target when required

With `auto`, the pristine target will be run when running
west with an existing build folder containing a build system and:

- Selecting a different board from the one currently in the build system
- Selecting a different application from the one currently in the build
  system

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

mbolivar commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -73,6 +73,13 @@ Additionally you can specify the build system target using the ``--target``
+using a different build dir::
+
+  west build qemu_x86 samples/philosophers
+  west build -p reel_board samples/hello_world
```
```suggestion
  west build -p -b reel_board samples/hello_world
```

mbolivar commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -73,6 +73,13 @@ Additionally you can specify the build system target using the ``--target``
+This is particularly handy if you want to switch source dirs or boards without
+using a different build dir::
+
+  west build qemu_x86 samples/philosophers
```
```suggestion
  west build -b qemu_x86 samples/philosophers
```

mbolivar commented on doc/guides/west/config.rst:
```diff
@@ -150,3 +150,22 @@ commands.
+-----------------------------------------------
+
+The following table documents configuration options supported by zephyr's
+extension commands (found in :file:`scripts/west_commands`)
```
```suggestion
extension commands (found in :file:`scripts/west_commands`).
```

mbolivar commented on doc/guides/west/config.rst:
```diff
@@ -150,3 +150,22 @@ commands.
+   * - ``build.auto-pristine``
+     - Boolean. If ``true`` (the default is ``false``), then ``west build`` will
+       automatically run the ``pristine`` target if a build system is present
+       and the build will fail otherwise (eg. the user has specified a different
```
```suggestion
       and the build will fail otherwise (e.g. the user has specified a different
```

mbolivar commented on doc/guides/west/config.rst:
```diff
@@ -150,3 +150,22 @@ commands.
+     - Description
+   * - ``build.auto-pristine``
+     - Boolean. If ``true`` (the default is ``false``), then ``west build`` will
+       automatically run the ``pristine`` target if a build system is present
```
```suggestion
       automatically run the ``pristine`` target before building, if a build system is present
```

mbolivar commented on doc/guides/west/config.rst:
```diff
@@ -150,3 +150,22 @@ commands.
+     - Boolean. If ``true`` (the default is ``false``), then ``west build`` will
+       automatically run the ``pristine`` target if a build system is present
+       and the build will fail otherwise (eg. the user has specified a different
+       board or application)
```
```suggestion
       board or application from the one previously used to make the build directory.)
```

mbolivar commented on scripts/west_commands/build.py:
```diff
@@ -291,19 +299,39 @@ def _sanity_check(self):
+            # Tricky coner-case: The user has not specified a build folder but
+            # there was one in the CMake cache. Since this is going to be
+            # invalidated, reset to CWD and re-run the basic tests.
+            if((boards_mismatched and not apps_mismatched) and
```
PEP8 nit, please add space after `if`

mbolivar reviewed (commented):
Nice patches! I tested this out and it's working the way I expect. So I'm +1 if the docs and PEP8 nit are resolved.

carlescufi commented:
Thanks for the review @mbolivar! I've changed the last commit so that now the config option takes the form:

```
[build]
pristine = {never, auto, always}
```

mbolivar reviewed (changes requested):
I really like the idea of always/auto/never, but I have some suggestions for how it could be improved.

carlescufi commented:
@mbolivar Thanks for the great tips! I really like your proposals and I have implemented them. Note that I have now squashed the two commits into a single one since they really belong together now that they are bound to the same behavior.

mbolivar commented on scripts/west_commands/build.py:
```diff
@@ -127,10 +143,29 @@ def do_run(self, args, remainder):
+                log.wrn('treating unknown build.pristine value "{}" as "never"'.
+                        format(pristine))
+                pristine = 'never'
+        self.auto_pristine = True if pristine == 'auto' else False
```
```suggestion
        self.auto_pristine = pristine == 'auto'
```

Thanks for renaming. This is also a bit redundant, since `==` will always evaluate to True or False.

You could also do `self.auto_pristine = (pristine == 'auto')` if that's more clear.

carlescufi replied:
Fixed! For some reason this did not work when I first wrote it and switched to the longer, explicit version. But now I re-tried and worked fine so I agree this is the correct way to do this.

mbolivar replied:
Great, looking very nice, thanks!

mbolivar reviewed (approved):
Looks good to me; thanks for making these changes! I have one last style nit, but I'm good with this as-is, +1.

overheat commented:
Works perfectly on my PC. Thanks!

dbkinder commented on scripts/west_commands/build.py:
```diff
@@ -106,6 +111,17 @@ def do_add_parser(self, parser_adder):
+                            help='''Control whether the pristine target is run
+                            before building if a build system is present in the
+                            build dir. --pristine is the same as
+                            --pristine=always. If set to auto, the pristine
```
No equal sign, right? ``--pristine always``

carlescufi replied:
both are correct, `--pristine always` and `--pristine=always`. I used the one with the equal sign to make it fool-proof.

dbkinder commented on doc/guides/west/config.rst:
```diff
@@ -150,3 +150,29 @@ commands.
+Zephyr Extension Commands Configuration Options
+-----------------------------------------------
+
+The following table documents configuration options supported by zephyr's
```
``Zephyr`` (upper case)

carlescufi replied:
Actually I intentionally used lower-case because I am referring to the repository, since extension commands are per-repository.

dbkinder reviewed (changes requested):
a couple of nits

dbkinder reviewed (approved):
oh, OK.
